proposals/0161-sbpf-feature-gating.md

```diff
@@ -0,0 +1,138 @@
+---
+simd: '0161'
+title: SBPF versioning and feature gating
+authors:
+  - Alexander Meißner
+  - Alessandro Decina
+category: Standard
+type: Core
+status: Draft
+created: 2024-07-15
+---
+
+## Summary
+
+An explicit versioning system for SBPF programs, that allows enabling or
+disabling features on a per-program basis.
+
+## Motivation
+
+SBPF has evolved over the years and will continue to evolve. Changes are
+necessary among other reasons to improve performance, address security concerns
+and provide new, better features to dapp developers.
+
+Today, the only way to introduce changes to the program runtime is via feature
+gates [link that explains what feature gates are]. For example we have used
+feature gates in the past to restrict new deployments of programs owned by
+deprecated loaders. Feature gates alone are not sufficient to evolve SBPF
+though - we are missing a mechanism to enable (or disable) specific features on
+a per-program basis.
+
+As an example, over two years ago we decided to change the way the stack works
+for SBPF programs - from allocating a fixed 4kB per function frame, to letting
+functions dynamically specify how much stack they need. This change was
+implemented in both the program runtime and the toolchain (LLVM), but as of
+today it has not yet been deployed, because it's essentially too hard to do so:
+upon executing a program, should the program runtime use the old or the new
+stack allocation strategy?
+
+We propose to introduce an explicit versioning scheme for SBPF programs:
+programs will encode which SBPF version they are compiled for; based on this
+version, the program runtime will alter its behavior. In order to prevent an
+"extension hell" it will not be possible to opt into specific changes via
+flags, instead an entire set of changes is reduced to a single SBPF version.
+
+## New Terminology
+
+SBPF version: the SBPF version number a program was compiled for.
+
+## Detailed Design
+
+Every program must signal which SBPF version it was compiled for via
+the `e_flags` field in the
+[ELF header](https://refspecs.linuxfoundation.org/elf/gabi4+/ch4.eheader.html).
+Block explorers are recommended to display this SBPF version field for program
+accounts.
+
+For each SBPF version two feature gates will control its availability:
```

Comment:
I'm not sure that we need such a strict system. I think we could probably have several versions active and deployable at the same time, without forcing people to update toolchain and redeploy or pay extra costs. 

Take the following examples:

- dynamic frames: from the vm perspective, it _is_ cheaper to use dynamic frames, because programs potentially page in less memory if an X-deep call stack fits in 1 page vs taking X pages because of fixed framing. We might want to push people to update to increase performance.

- the `sub imm` operand swap change: this doesn't materially change anything. It's a nice thing for us to fix for aesthetic reasons, but it doesn't really change anything for developers. 

I think I'd prefer an ad-hoc system where depending on the impact of a new SBPF version, we decide whether and how to rug earlier versions (in practice I expect this to happen very rarely).

Reply:
Mandating that all deployments are targeting the newest version does not force anybody to redeploy. It only means that dapp devs have to keep their toolchain updated when they re/deploy. You are right in that it is not necessary to be this strict, but it has the nice effect of allowing us to get more testing exposure for the newest toolchain version.

Reply:
I agree that we shouldn't force re-deployments. Mandating that when you do deploy, you can only deploy the latest SBPF version is a reasonable tradeoff between usability and actually getting new versions adopted.